src/masternodes/order.h

```diff
@@ -0,0 +1,222 @@
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(ownerAddress);
+        READWRITE(tokenFrom);
+        READWRITE(tokenTo);
+        READWRITE(VARINT(idTokenFrom.v));
+        READWRITE(VARINT(idTokenTo.v));
+        READWRITE(amountFrom);
+        READWRITE(orderPrice);
+        READWRITE(expiry);
+    }
+};
+
+class COrderImplemetation : public COrder
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    uint256 closeTx;
+    uint32_t creationHeight; 
+    uint32_t closeHeight;
+
+    COrderImplemetation()
+        : COrder()
+        , creationTx()
+        , closeTx()
+        , creationHeight(-1)
+        , closeHeight(-1)
+    {}
+    ~COrderImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(COrder, *this);
+        READWRITE(creationTx);
+        READWRITE(closeTx);
+        READWRITE(creationHeight);
+        READWRITE(closeHeight);
+    }
+};
+
+class CFulfillOrder
+{
+public:
+    //! basic properties
+    std::string ownerAddress;
+    uint256 orderTx;
+    CAmount amount;
+
+    CFulfillOrder()
+        : ownerAddress("")
+        , orderTx(uint256())
+        , amount(0)
+    {}
+    virtual ~CFulfillOrder() = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(ownerAddress);
+        READWRITE(orderTx);
+        READWRITE(amount);
+    }
+};
+
+class CFulfillOrderImplemetation : public CFulfillOrder
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    uint256 closeTx;
+    uint32_t creationHeight; 
+    uint32_t closeHeight;
+
+    CFulfillOrderImplemetation()
+        : CFulfillOrder()
+        , creationTx()
+        , closeTx()
+        , creationHeight(-1)
+        , closeHeight(-1)
+    {}
+    ~CFulfillOrderImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(CFulfillOrder, *this);
+        READWRITE(creationTx);
+        READWRITE(closeTx);
+        READWRITE(creationHeight);
+        READWRITE(closeHeight);
+    }
+};
+
+class CCloseOrder
+{
+public:
+    //! basic properties
+    uint256 orderTx;
+
+    CCloseOrder()
+        : orderTx(uint256())
+    {}
+    virtual ~CCloseOrder() = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(orderTx);
+    }
+};
+
+class CCloseOrderImplemetation : public CCloseOrder
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    uint32_t creationHeight; 
```

Comment:
These should be `closeTx` and `closeHeight`?

Reply:
My logic (as I took it from tokens) was for every tx type to have its creationTx id, basically txid. So for close order tx which is also a transaction on chain I store its txid in db. And in COrder there is creationTx and also closeTx which refers on this close tx.